File: src/ableton_mcp/tools/devices.py
"""Device listing and parameter control."""

from typing import Literal

from ..connection import get_connection

TrackType = Literal["track", "return", "master"]


def register(mcp):
    @mcp.tool()
    def list_devices(track_index: int = 0, track_type: TrackType = "track") -> dict:
        """List a track's devices, including rack chains and drum pads. Each
        device has a LOM path usable with live_get/live_set/live_call for
        anything the curated tools don't cover (e.g. devices nested inside
        racks)."""
        return get_connection().request(
            "get_devices", {"track_index": track_index, "track_type": track_type}
        )

    @mcp.tool()
    def device_parameters(
        track_index: int, device_index: int, track_type: TrackType = "track"
    ) -> dict:
        """List a device's parameters: name, value, range, display string, and
        (for switches/menus) the list of valid options."""
        return get_connection().request(
            "get_device_parameters",
            {"track_index": track_index, "track_type": track_type,
             "device_index": device_index},
        )

    @mcp.tool()
    def set_device_parameter(
        track_index: int,
        device_index: int,
        parameter: int | str,
        value: float | str,
        track_type: TrackType = "track",
    ) -> dict:
        """Set a device parameter by index or name. Pass a number for
        continuous parameters (clamped to the parameter's range) or an option
        name (e.g. "High") for quantized ones. "Device On" with 0/1 turns the
        device off/on."""
        return get_connection().request(
            "set_device_parameter",
            {"track_index": track_index, "track_type": track_type,
             "device_index": device_index, "parameter": parameter, "value": value},
        )

    @mcp.tool()
    def delete_device(
        track_index: int, device_index: int, track_type: TrackType = "track"
    ) -> dict:
        """Remove a device from a track."""
        return get_connection().request(
            "delete_device",
            {"track_index": track_index, "track_type": track_type,
             "device_index": device_index},
        )

File: server/src/tools/devices.ts
// Device listing and parameter control.

import * as z from "zod";
import { bridge, type Registrar } from "./helpers.js";

const trackType = z.enum(["track", "return", "master"]).default("track");

export const register: Registrar = (server) => {
  server.registerTool(
    "list_devices",
    {
      description:
        "List a track's devices, including rack chains and drum pads. Each " +
        "device has a LOM path usable with live_get/live_set/live_call for " +
        "anything the curated tools don't cover (e.g. devices nested " +
        "inside racks).",
      inputSchema: z.object({
        track_index: z.number().int().default(0),
        track_type: trackType,
      }),
    },
    async (args) => bridge("get_devices", args),
  );

  server.registerTool(
    "device_parameters",
    {
      description:
        "List a device's parameters: name, value, range, display string, " +
        "and (for switches/menus) the list of valid options.",
      inputSchema: z.object({
        track_index: z.number().int(),
        device_index: z.number().int(),
        track_type: trackType,
      }),
    },
    async (args) => bridge("get_device_parameters", args),
  );

  server.registerTool(
    "set_device_parameter",
    {
      description:
        "Set a device parameter by index or name. Pass a number for " +
        "continuous parameters (clamped to the parameter's range) or an " +
        'option name (e.g. "High") for quantized ones. "Device On" with ' +
        "0/1 turns the device off/on.",
      inputSchema: z.object({
        track_index: z.number().int(),
        device_index: z.number().int(),
        parameter: z.union([z.number().int(), z.string()]),
        value: z.union([z.number(), z.string()]),
        track_type: trackType,
      }),
    },
    async (args) => bridge("set_device_parameter", args),
  );

  server.registerTool(
    "delete_device",
    {
      description: "Remove a device from a track.",
      inputSchema: z.object({
        track_index: z.number().int(),
        device_index: z.number().int(),
        track_type: trackType,
      }),
    },
    async (args) => bridge("delete_device", args),
  );
};
